File: studio/events.py
"""SSE event stream for run journals."""

import asyncio
import json
import os

from harness import packets

TERMINAL_EVENTS = frozenset(
    {
        "shipped",
        "blocked",
        "escalated",
        "rejected_by_user",
        "killed",
        "failed",
        "paused",
    }
)


def _format_sse(event):
    data = json.dumps(event)
    seq = event.get("seq", "")
    ev = event.get("event", "message")
    return f"id: {seq}\nevent: {ev}\ndata: {data}\n\n"


async def stream_events(run_dir, run_data, last_event_id=-1):
    """Async generator yielding SSE messages for a run."""
    journal_path = os.path.join(run_dir, "journal.jsonl")
    last_id = int(last_event_id)
    sent = set()

    def new_events():
        events = packets.journal_read(journal_path)
        out = []
        for e in events:
            if "seq" not in e:
                continue
            if e["seq"] > last_id and e["seq"] not in sent:
                out.append(e)
        return out

    for e in new_events():
        sent.add(e["seq"])
        last_id = max(last_id, e["seq"])
        yield _format_sse(e)

    poll_s = float(os.environ.get("STUDIO_SSE_POLL_SEC", "0.25"))
    while True:
        found = False
        for e in new_events():
            found = True
            sent.add(e["seq"])
            last_id = max(last_id, e["seq"])
            yield _format_sse(e)
        state = run_data.get("state", "")
        terminal = state in (
            "shipped",
            "blocked",
            "escalated",
            "rejected",
            "killed",
            "failed",
        )
        if terminal and not found:
            events = packets.journal_read(journal_path)
            pending = [
                e
                for e in events
                if "seq" in e and e["seq"] > last_id and e["seq"] not in sent
            ]
            for e in pending:
                sent.add(e["seq"])
                last_id = max(last_id, e["seq"])
                yield _format_sse(e)
            yield "event: done\ndata: {}\n\n"
            break
        if terminal:
            drained = True
            events = packets.journal_read(journal_path)
            for e in events:
                if "seq" in e and e["seq"] > last_id:
                    drained = False
                    break
            if drained:
                yield "event: done\ndata: {}\n\n"
                break
        await asyncio.sleep(poll_s)
        path = os.path.join(run_dir, "run.json")
        if os.path.exists(path):
            with open(path, encoding="utf-8") as f:
                run_data.update(json.load(f))

File: src/server/events.ts
/** Port of backend/studio/events.py — SSE stream for run journals. */
import fs from "node:fs";
import path from "node:path";
import { journalRead, type JournalEvent } from "./packets";

const TERMINAL_STATES = new Set([
  "shipped",
  "blocked",
  "escalated",
  "rejected",
  "killed",
  "failed",
]);

function formatSse(event: JournalEvent): string {
  const data = JSON.stringify(event);
  const seq = event.seq ?? "";
  const ev = event.event ?? "message";
  return `id: ${seq}\nevent: ${ev}\ndata: ${data}\n\n`;
}

export async function* streamEvents(
  runDir: string,
  runData: Record<string, unknown>,
  lastEventId = -1,
): AsyncGenerator<string> {
  const journalPath = path.join(runDir, "journal.jsonl");
  let lastId = lastEventId;
  const sent = new Set<number>();

  const newEvents = (): JournalEvent[] => {
    const events = journalRead(journalPath);
    const out: JournalEvent[] = [];
    for (const e of events) {
      if (typeof e.seq !== "number") continue;
      if (e.seq > lastId && !sent.has(e.seq)) out.push(e);
    }
    return out;
  };

  for (const e of newEvents()) {
    sent.add(e.seq!);
    lastId = Math.max(lastId, e.seq!);
    yield formatSse(e);
  }

  const pollS = parseFloat(process.env.STUDIO_SSE_POLL_SEC ?? "0.25");
  while (true) {
    let found = false;
    for (const e of newEvents()) {
      found = true;
      sent.add(e.seq!);
      lastId = Math.max(lastId, e.seq!);
      yield formatSse(e);
    }

    const state = String(runData.state ?? "");
    const terminal = TERMINAL_STATES.has(state);

    if (terminal && !found) {
      for (const e of journalRead(journalPath)) {
        if (typeof e.seq === "number" && e.seq > lastId && !sent.has(e.seq)) {
          sent.add(e.seq);
          lastId = Math.max(lastId, e.seq);
          yield formatSse(e);
        }
      }
      yield "event: done\ndata: {}\n\n";
      break;
    }

    if (terminal) {
      let drained = true;
      for (const e of journalRead(journalPath)) {
        if (typeof e.seq === "number" && e.seq > lastId) {
          drained = false;
          break;
        }
      }
      if (drained) {
        yield "event: done\ndata: {}\n\n";
        break;
      }
    }

    await new Promise((r) => setTimeout(r, pollS * 1000));
    const runJson = path.join(runDir, "run.json");
    if (fs.existsSync(runJson)) {
      Object.assign(runData, JSON.parse(fs.readFileSync(runJson, "utf-8")));
    }
  }
}
